File: src/wacc.py
"""
WACC builder.

Per dcf/SPEC_methodology section 5:
  1. Unlever each peer beta:  Bu = Be / (1 + (1 - t) × D/E)
  2. Take median (or mean) unlevered beta of peer set
  3. Re-lever to target capital structure:  Be_target = Bu_med × (1 + (1 - t) × D/E_target)
  4. CAPM cost of equity:  Ke = Rf + Be × ERP
  5. After-tax cost of debt:  Kd_at = Kd × (1 - t)
  6. WACC = We × Ke + Wd × Kd_at
"""
import logging
import statistics

logger = logging.getLogger(__name__)


def _unlever_beta(levered_beta: float, de_ratio: float, tax_rate: float) -> float:
    """Bu = Be / (1 + (1 - t) × D/E)"""
    denom = 1 + (1 - tax_rate) * de_ratio
    return levered_beta / denom if denom else levered_beta


def _relever_beta(unlevered_beta: float, target_de: float, target_tax: float) -> float:
    """Be_target = Bu × (1 + (1 - t) × D/E_target)"""
    return unlevered_beta * (1 + (1 - target_tax) * target_de)


def compute_wacc(
    peer_set,                                 # PeerSet
    target_market_cap: float,                 # $M
    target_debt: float,                       # $M (long-term debt last historical)
    risk_free_rate: float,
    equity_risk_premium: float,
    cost_of_debt_pretax: float,
    target_tax_rate: float = 0.21,
    target_de_ratio: float | None = None,
    fallback_beta: float = 1.0,
    sector: str = "standard",
    ledger=None,
):
    """
    Returns WACCOutput with full peer-set unlever/relever + CAPM build-up.

    Falls back to single-ticker beta when peer_set has no peers (LLM unavailable).
    """
    from schemas.financial_data import WACCOutput

    from src.assumption_registry import resolve as _resolve

    if peer_set.peers:
        unlevered_betas = [
            _unlever_beta(p.levered_beta, p.de_ratio, p.tax_rate)
            for p in peer_set.peers
        ]
        median_bu = statistics.median(unlevered_betas)
        if ledger is not None:
            ledger.record_derived(
                "wacc", "median_unlevered_beta", None, value=round(median_bu, 4),
                formula="median(unlever(peer.beta))",
                inputs=[("peers", p.ticker, None) for p in peer_set.peers],
            )
    else:
        a = _resolve("sector_beta", sector=sector)
        median_bu = a.value if a else fallback_beta
        logger.warning("No peers in set — using sector-median beta %.2f", median_bu)
        if ledger is not None:
            if a:
                ledger.record_assumption("wacc", "median_unlevered_beta", None,
                                         value=round(median_bu, 4),
                                         rationale=a.rationale, basis=a.basis)
            else:
                ledger.record_unverified("wacc", "median_unlevered_beta", None,
                                         value=round(median_bu, 4),
                                         reason="no peers and no sector beta declared")

    de = target_de_ratio if target_de_ratio is not None else peer_set.target_de_ratio
    target_levered_beta = _relever_beta(median_bu, de, target_tax_rate)

    # CAPM
    cost_of_equity = risk_free_rate + target_levered_beta * equity_risk_premium

    # After-tax cost of debt
    after_tax_kd = cost_of_debt_pretax * (1 - target_tax_rate)

    # Capital structure weights — use market values
    total_capital = target_market_cap + target_debt
    if total_capital <= 0:
        equity_weight = 1.0
        debt_weight = 0.0
    else:
        equity_weight = target_market_cap / total_capital
        debt_weight   = target_debt / total_capital

    # WACC
    wacc = equity_weight * cost_of_equity + debt_weight * after_tax_kd
    # Sanity clamp
    wacc = max(0.05, min(wacc, 0.30))

    if ledger is not None:
        ledger.record_derived(
            "wacc", "wacc", None, value=round(wacc, 4),
            formula="We*Ke + Wd*Kd*(1-t)",
            inputs=[("wacc", "cost_of_equity", None), ("wacc", "after_tax_cost_of_debt", None)],
        )

    return WACCOutput(
        peers=peer_set.peers,
        median_unlevered_beta=round(median_bu, 4),
        target_levered_beta=round(target_levered_beta, 4),
        target_de_ratio=de,
        risk_free_rate=risk_free_rate,
        equity_risk_premium=equity_risk_premium,
        cost_of_equity=round(cost_of_equity, 4),
        cost_of_debt_pretax=cost_of_debt_pretax,
        tax_rate=target_tax_rate,
        after_tax_cost_of_debt=round(after_tax_kd, 4),
        target_market_cap=target_market_cap,
        target_debt=target_debt,
        target_total_capital=total_capital,
        equity_weight=round(equity_weight, 4),
        debt_weight=round(debt_weight, 4),
        wacc=round(wacc, 4),
    )

File: finmodel-core/fm-value/src/wacc.rs
//! WACC builder — port of `src/wacc.py`.

use crate::types::{PeerSet, WACCOutput};

/// CAPM cost of equity: Re = Rf + beta × ERP.
pub fn cost_of_equity(rf: f64, erp: f64, beta: f64) -> f64 {
    rf + beta * erp
}

/// Bu = Be / (1 + (1 − t) × D/E)
pub fn unlever_beta(levered_beta: f64, de_ratio: f64, tax_rate: f64) -> f64 {
    let denom = 1.0 + (1.0 - tax_rate) * de_ratio;
    if denom == 0.0 {
        levered_beta
    } else {
        levered_beta / denom
    }
}

/// Be_target = Bu × (1 + (1 − t) × D/E_target)
pub fn relever_beta(unlevered_beta: f64, target_de: f64, target_tax: f64) -> f64 {
    unlevered_beta * (1.0 + (1.0 - target_tax) * target_de)
}

fn median(mut xs: Vec<f64>) -> f64 {
    if xs.is_empty() {
        return 0.0;
    }
    xs.sort_by(|a, b| a.partial_cmp(b).unwrap_or(std::cmp::Ordering::Equal));
    let n = xs.len();
    if n % 2 == 1 {
        xs[n / 2]
    } else {
        (xs[n / 2 - 1] + xs[n / 2]) / 2.0
    }
}

/// Lightweight WACC from explicit capital-structure inputs (unit-test helper).
pub fn calculate(input: &WACCInput) -> f64 {
    let re = cost_of_equity(input.risk_free_rate, input.equity_risk_premium, input.beta);
    input.equity_pct * re + input.debt_pct * input.cost_of_debt * (1.0 - input.tax_rate)
}

#[derive(Debug, Clone)]
pub struct WACCInput {
    pub risk_free_rate: f64,
    pub equity_risk_premium: f64,
    pub beta: f64,
    pub cost_of_debt: f64,
    pub tax_rate: f64,
    pub debt_pct: f64,
    pub equity_pct: f64,
}

/// Full peer-set unlever/relever + CAPM WACC. Mirrors `compute_wacc`.
pub fn compute_wacc(
    peer_set: &PeerSet,
    target_market_cap: f64,
    target_debt: f64,
    risk_free_rate: f64,
    equity_risk_premium: f64,
    cost_of_debt_pretax: f64,
    target_tax_rate: f64,
    target_de_ratio: Option<f64>,
    fallback_beta: f64,
) -> WACCOutput {
    let median_bu = if peer_set.peers.is_empty() {
        fallback_beta
    } else {
        median(
            peer_set
                .peers
                .iter()
                .map(|p| unlever_beta(p.levered_beta, p.de_ratio, p.tax_rate))
                .collect(),
        )
    };
    let de = target_de_ratio.unwrap_or(peer_set.target_de_ratio);
    let target_levered_beta = relever_beta(median_bu, de, target_tax_rate);
    let ke = cost_of_equity(risk_free_rate, equity_risk_premium, target_levered_beta);
    let after_tax_kd = cost_of_debt_pretax * (1.0 - target_tax_rate);
    let total_capital = target_market_cap + target_debt;
    let (equity_weight, debt_weight) = if total_capital <= 0.0 {
        (1.0, 0.0)
    } else {
        (target_market_cap / total_capital, target_debt / total_capital)
    };
    let mut wacc = equity_weight * ke + debt_weight * after_tax_kd;
    wacc = wacc.clamp(0.05, 0.30);

    WACCOutput {
        peers: peer_set.peers.clone(),
        median_unlevered_beta: round4(median_bu),
        target_levered_beta: round4(target_levered_beta),
        target_de_ratio: de,
        risk_free_rate,
        equity_risk_premium,
        cost_of_equity: round4(ke),
        cost_of_debt_pretax,
        tax_rate: target_tax_rate,
        after_tax_cost_of_debt: round4(after_tax_kd),
        target_market_cap,
        target_debt,
        target_total_capital: total_capital,
        equity_weight: round4(equity_weight),
        debt_weight: round4(debt_weight),
        wacc: round4(wacc),
    }
}

/// Empty-peer fallback PeerSet used when no LLM/market peers are available.
pub fn fallback_peer_set(ticker: &str, market_cap: f64, de: f64) -> PeerSet {
    PeerSet {
        target_ticker: ticker.to_string(),
        target_market_cap: market_cap,
        target_de_ratio: de,
        peers: Vec::new(),
        excluded: Vec::new(),
        source: "fallback".into(),
    }
}

fn round4(x: f64) -> f64 {
    (x * 10_000.0).round() / 10_000.0
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_wacc_known_value() {
        let input = WACCInput {
            risk_free_rate: 0.03,
            equity_risk_premium: 0.05,
            beta: 1.2,
            cost_of_debt: 0.04,
            tax_rate: 0.25,
            debt_pct: 0.4,
            equity_pct: 0.6,
        };
        let re = cost_of_equity(input.risk_free_rate, input.equity_risk_premium, input.beta);
        assert!((re - 0.09).abs() < 1e-12);
        let wacc = calculate(&input);
        assert!((wacc - 0.066).abs() < 1e-12);
    }

    #[test]
    fn test_compute_wacc_no_peers_clamped() {
        let ps = fallback_peer_set("TEST", 1000.0, 0.3);
        let out = compute_wacc(&ps, 1000.0, 300.0, 0.04, 0.05, 0.04, 0.21, Some(0.3), 1.0);
        assert!(out.wacc >= 0.05 && out.wacc <= 0.30);
        assert!((out.median_unlevered_beta - 1.0).abs() < 1e-9);
        assert!(out.equity_weight > 0.0);
    }

    #[test]
    fn test_unlever_relever_roundtrip() {
        let bl = 1.2;
        let de = 0.5;
        let t = 0.25;
        let bu = unlever_beta(bl, de, t);
        let back = relever_beta(bu, de, t);
        assert!((back - bl).abs() < 1e-12);
    }
}
